kani-compiler/src/kani_middle/attributes.rs

```diff
@@ -989,3 +991,54 @@ pub fn matches_diagnostic<T: CrateDef>(tcx: TyCtxt, def: T, attr_name: &str) ->
     }
     false
 }
+
+/// Parse an attribute using `syn`.
+///
+/// This provides a user-friendly interface to manipulate than the internal compiler AST.
+fn syn_attr(attr: &Attribute) -> syn::Attribute {
+    let attr_str = rustc_ast_pretty::pprust::attribute_to_string(attr);
+    let parser = syn::Attribute::parse_outer;
+    parser.parse_str(&attr_str).unwrap().pop().unwrap()
+}
+
+/// Return a more user-friendly string for path by trying to remove unneeded whitespace.
+///
+/// `quote!()` and `TokenString::to_string()` introduce unnecessary space around separators.
+/// This happens because these methods end up using TokenStream display, which has no
+/// guarantees on the format printed.
+/// <https://doc.rust-lang.org/proc_macro/struct.TokenStream.html#impl-Display-for-TokenStream>
+///
+/// E.g.: The path `<[char; 10]>::foo` printed with token stream becomes `< [ char ; 10 ] > :: foo`.
+/// while this function turns this into `<[char ; 10]>::foo`.
+///
+/// Thus, this can still be improved to handle the `qself.ty`.
+///
+/// We also don't handle path segments, but users shouldn't pass generic arguments to our
+/// attributes.
+fn pretty_type_path(path: &TypePath) -> String {
+    fn segments_str<'a, I>(segments: I) -> String
+    where
+        I: IntoIterator<Item = &'a PathSegment>,
+    {
+        // We don't bother with path arguments for now since users shouldn't provide them.
+        segments
+            .into_iter()
+            .map(|segment| segment.to_token_stream().to_string())
+            .intersperse("::".to_string())
+            .collect()
```

Comment:
Wouldn't be better to use a String buffer here? We can replace `intersperse` with a `peekable` iterator to avoid creating an intermediate collection

Reply:
Maybe the way you choose is more readable since we won't need loops. Is that why?

Reply:
I thought it was more readable, and this code is only invoked for error messages, which shouldn't occur a lot, thus performance is not a concern. Let me know if you want me to make an explicit loop.

Reply:
Not a problem! If performance is not a concern, we should focus on readability and the way it is is fine.